File: baekjoon/basic2/17298/17298.py
# 백준 17298 : 오큰수 (https://www.acmicpc.net/problem/17298)
import sys

n = int(sys.stdin.readline())

a = list(map(int, sys.stdin.readline().split()))

stk = [0]
ret = [-1] * 1000004
for i, v in enumerate(a):
    if i == 0:
        continue
    while stk and a[stk[-1]] < a[i]:
        ret[stk[-1]] = a[i]
        stk.pop()
    stk.append(i)

for i in range(n):
    sys.stdout.write(str(ret[i])+" ")
sys.stdout.write("\n")

File: baekjoon/basic2/17298/17298.cpp
// 백준 : 17298 오큰수 (https://www.acmicpc.net/problem/17298)
#include <iostream>
#include <vector>
#include <stack>
using namespace std;
int n, tmp;
const int N = 1000004;
vector<int> a(N), ret(N, -1);
stack<int> stk;

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    cin >> n;
    for(int i=0; i<n; i++){
        cin >> a[i];
        while(!stk.empty() && a[stk.top()] < a[i]){
            ret[stk.top()] = a[i];
            stk.pop();
        }
        stk.push(i);
    }
    for(int i=0; i<n; i++){
        cout << ret[i] << " ";
    }
    cout << "\n";

    return 0;
}